src/caller.jl

```diff
@@ -0,0 +1,189 @@
+end
+
+function prepare_argument(x::Int64)
+    return Any[ mapping_types_reversed[:INT_CMD], Ptr{Cvoid}(x) ], nothing
+end
+
+function prepare_argument(x::String)
+    return Any[ mapping_types_reversed[:STRING_CMD], libSingular.copy_string_to_void(x) ], nothing
+end
+
+function prepare_argument(x::PolyRing)
+    new_ptr = libSingular.get_ring_ref(x.ptr)
+    return Any[ mapping_types_reversed[:RING_CMD], new_ptr ], x
+end
+
+function prepare_argument(x::spoly)
+    rng = parent(x)
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:POLY_CMD], libSingular.copy_polyptr_to_void(x.ptr, rng_ptr) ], rng
+end
+
+function prepare_argument(x::svector)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:VECTOR_CMD], libSingular.copy_polyptr_to_void(x.ptr, rng_ptr) ], rng
+end
+
+function prepare_argument(x::sideal)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:IDEAL_CMD], libSingular.copy_idealptr_to_void(x.ptr, rng_ptr)], rng
+end
+
+function prepare_argument(x::smodule)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:MODUL_CMD], libSingular.copy_idealptr_to_void(x.ptr, rng_ptr)], rng
+end
+
+function prepare_argument(x::sresolution)
+    rng = base_ring(x)
+    res = Any[ mapping_types_reversed[:RESOLUTION_CMD], libSingular.create_syStrategy_data(x.ptr, rng.ptr) ]
+    return res, rng
+end
+
+function prepare_argument(x::Any)
+    if x.ptr isa libSingular.number
+        ptr = x.ptr
+        rng = parent(x)
+        new_ptr = libSingular.n_Copy(ptr, rng.ptr)
+        return Any[ mapping_types_reversed[:NUMBER_CMD], new_ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.ip_smatrix
+        rng = parent(x)
+        return Any[ mapping_types_reversed[:MATRIX_CMD], libSingular.mpCopy(x.ptr, rng.ptr).cpp_object ], rng
+    elseif x.ptr isa libSingular.__mpz_struct
+        return Any[ mapping_types_reversed[:BIGINT_CMD], x.ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.sip_smap
+        return Any[ mapping_types_reversed[:MAP_CMD], x.ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.bigintmat
+        return Any[ mapping_types_reversed[:BIGINTMAT_CMD], x.ptr.cpp_object ], nothing
+    end
+
+end
+
+function low_level_caller_rng(lib::String, name::String, ring, args...)
+    libSingular.load_library(lib)
+    arguments = [prepare_argument(i) for i in args]
+    arguments = Any[ i for (i, j) in arguments ]
+    return_value = libSingular.call_singular_library_procedure(name, ring.ptr, arguments)
+    return convert_return_list(return_value, ring)
+end
+
+function low_level_caller(lib::String, name::String, args...)
+    libSingular.load_library(lib)
+    arguments = [prepare_argument(i) for i in args]
+    rng = nothing
+    for (i, j) in arguments
+        if j != nothing
+            rng = j
```

Comment:
So, you take the last ring of all the rings. Why not the first (and then insert a `break` here)? Or, even better: Why not check that all subsequent rings are compatible with the first -- as I also suggested for other variant of this function. With pseudo code like this:
```julia
arguments = []
for i in args
  i, j = prepare_argument(i)
  push!(arguments, i)
  if j != nothing
    if rng == nothing
      rng = j
    else
      # TODO: check "compatibility, e.g. rng.ptr == j.ptr
      # (but perhaps that is too restrictive?)
    end
  end
```
With that, the two function variants can be merged, with this ring-free variant becoming just `return low_level_caller(lib, name, nothing, args...)`